lib/pure/asyncnet.nim

```diff
@@ -822,40 +822,47 @@ proc sendTo*(socket: AsyncSocket, address: string, port: Port, data: string,
     else:
       raise newException(IOError, "Couldn't resolve address: " & address)
 
-proc recvFrom*(socket: AsyncSocket, size: int,
-               flags = {SocketFlag.SafeDisconn}):
-              owned(Future[tuple[data: string, address: string, port: Port]])
+proc recvFrom*(socket: AsyncSocket, data: FutureVar[string], size: int,
+               address: FutureVar[string], port: FutureVar[Port],
+               flags = {SocketFlag.SafeDisconn}): owned(Future[int])
               {.async, since: (1, 3).} =
-  ## Receives a datagram data from ``socket``, which must be at least of size
-  ## ``size``. Returned future will complete once one datagram has been received
-  ## and will return tuple with: data of packet received; and address and port
-  ## of datagram's sender.
-  ## 
+  ## Receives a datagram data from ``socket`` into ``data``, which must be at
+  ## least of size ``size``. The address and port of datagram's sender will be
+  ## stored into ``address`` and ``port``, respectively. Returned future will
+  ## complete once one datagram has been received, and will return size of
+  ## packet received.
+  ##
   ## If an error occurs an OSError exception will be raised.
   ## 
   ## This proc is normally used with connectionless sockets (UDP sockets).
+  ## 
+  ## **Notes**
+  ## * ``data`` must be initialized to the length of ``size``.
+  ## * ``address`` must be initialized to 39 in length.
   template adaptRecvFromToDomain(domain: Domain) =
     var lAddr = sizeof(sAddr).SockLen
     
-    let fut = await recvFromInto(AsyncFD(getFd(socket)), cstring(data), size,
-                                 cast[ptr SockAddr](addr sAddr), addr lAddr,
-                                 flags)
+    result = await recvFromInto(AsyncFD(getFd(socket)), cstring(data.mget()), size,
+                                cast[ptr SockAddr](addr sAddr), addr lAddr,
+                                flags)
     
-    data.setLen(fut)
+    data.mget().setLen(result)
+    data.complete()
     
-    result.data = data
-    result.address = getAddrString(cast[ptr SockAddr](addr sAddr))
+    address.complete(getAddrString(cast[ptr SockAddr](addr sAddr)))
```

Comment:
This is still wrong, you need  `getAddrString` variant that will write to a `var` argument (in the same way that recvFromInto does with the `data.mget()` argument). Here you are still copying the string.

Reply:
@dom96 , I don't know if it's the best way, but I've been thinking about doing it, because it takes advantage of the pre-allocated memory:
```nim
    let addressTmp = getAddrString(cast[ptr SockAddr](addr sAddr))

    for i in 0 ..< len(addressTmp):
      address.mget()[i] = addressTmp[i]

    address.mget().setLen(len(addressTmp))    
    address.complete()
```
Because, in my view, it is safer than `moveMem()` and `copyMem()`, since the user may not correctly pre-allocate `address`.

Creating a new `getAddrString()` would fall into the same problem as `address = getAddrString()`, unless something is missed from me.

Reply:
Why don't you create a `getAddrString` like: `proc getAddrString(sAddr: ptr SockAddr, out: var string)`? That will solve the problem here.

Reply:
@dom96, sorry for my ignorance. I was able to understand and see how it works, now with more time.

Now there is a question as to how this should be done. Make a new `proc getAddrString(sockAddr: ptr SockAddr, strAddress: var string)` in the `asyncnet` package or `nativesockets`? Should it be `asyncnet`, should it be marked as exportable? Should another PR be done to first add this new `getAddrString`?

Also, there is the possibility, which I already did for testing, to use `winlean.inet_ntop()` or `posix.inet_ntop()` directly in the body of `recvFrom()`.

What better designer solution?

I also checked that the pre-allocation size of the `address FutureVar[string]` must be 46.

Reply:
I would put it in `nativesockets`, if you want to export it then it definitely shouldn't live in `asyncnet`.

> Also, there is the possibility, which I already did for testing, to use winlean.inet_ntop() or posix.inet_ntop() directly in the body of recvFrom().

I guess you can do this, but won't you just be redoing what getAddrString already implements? Might as well do it in its own function and make that available for others.

Reply:
I made the new `getAddrString` using the existing one as a base. However, I used `inet_ntop()` for both IPv4 and IPv6.

I tested on Windows 10 and Debian 10.